engine/core/src/main/java/org/datacleaner/job/builder/ComponentBuilderTransformerChangeListener.java

```diff
@@ -0,0 +1,100 @@
+import org.datacleaner.util.ReflectionUtils;
+
+public class ComponentBuilderTransformerChangeListener implements TransformerChangeListener {
+    private final ComponentBuilder _componentBuilder;
+    private final ConfiguredPropertyDescriptor _propertyDescriptor;
+
+    public ComponentBuilderTransformerChangeListener(ComponentBuilder componentBuilder,
+            ConfiguredPropertyDescriptor propertyDescriptor) {
+        _componentBuilder = componentBuilder;
+        _propertyDescriptor = propertyDescriptor;
+    }
+
+    @Override
+    public void onAdd(TransformerComponentBuilder<?> builder) {
+        // Do nothing.
+    }
+
+    @Override
+    public void onConfigurationChanged(TransformerComponentBuilder<?> builder) {
+        synchronizeInputColumns(builder);
+    }
+
+    @Override
+    public void onRequirementChanged(TransformerComponentBuilder<?> builder) {
+        // Do nothing.
+    }
+
+    @Override
+    public void onRemove(TransformerComponentBuilder<?> componentBuilder) {
+        synchronizeInputColumns(componentBuilder);
+    }
+
+    @Override
+    public void onOutputChanged(TransformerComponentBuilder<?> transformerJobBuilder,
+            List<MutableInputColumn<?>> outputColumns) {
+        synchronizeInputColumns(transformerJobBuilder);
+    }
+
+    private void synchronizeInputColumns(TransformerComponentBuilder<?> changedBuilder) {
+        if (!changedBuilder.equals(_componentBuilder)) {
+            final List<InputColumn<?>> availableColumns = _componentBuilder.getAnalysisJobBuilder()
+                    .getAvailableInputColumns(_componentBuilder, _propertyDescriptor.getTypeArgument(0));
+
+            for (ConfiguredPropertyDescriptor propertyDescriptor : _componentBuilder.getDescriptor()
+                    .getConfiguredPropertiesForInput()) {
+                final Object configuredProperty = _componentBuilder.getConfiguredProperty(propertyDescriptor);
+
+                if (configuredProperty != null) {
+                    if (ReflectionUtils.isInputColumn(configuredProperty.getClass())) {
+                        if (ReflectionUtils.isArray(configuredProperty)) {
+                            for (InputColumn<?> column : (InputColumn<?>[]) configuredProperty) {
+                                if (!availableColumns.contains(column)) {
+                                    _componentBuilder.removeInputColumn(column, propertyDescriptor);
+                                }
+                            }
+                        } else {
+                            if (!availableColumns.contains(configuredProperty)) {
+                                _componentBuilder.removeInputColumn((InputColumn<?>) configuredProperty,
+                                        propertyDescriptor);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public boolean equals(Object object) {
```

Comment:
`hashcode()` seems to be missing (`equals()` and `hashcode()` should always be overridden in pairs).

You can use `Objects.hash(_componentBuilder, _propertyDescriptor)` for a quick implementation.